Implement Group Score Update on match_score points change

Malaber opened the pull request.

Malaber committed:
Move self methods into self block

Malaber committed:
Update group_scores on match_score change if match is a group match

Malaber committed:
Implement Methods returning points of teams per match

These methods return group_points scored_points and received_points
respectively when given a team that is present in the match
They return 0 when given a team that isn't contesting in that match

Malaber committed:
Fix typo in received_points

Malaber committed:
Correctly handle matches that are not started yet

Malaber committed:
Implement update_group_score method

This method calculates all the group scores of the group given to it

Malaber committed:
Add method to check if match_score is part of a group match

Malaber committed:
Add GroupScore objects for every team to the group stage

Malaber committed:
Add group_scores to groups factory, one for every team

Malaber committed:
Make Match Factory configurable for group_stage and group factory

Malaber committed:
Remove unnecessary :group_stage_only tournament

Malaber committed:
Test if match_score point change triggers group score calculation

Malaber committed:
Dry out code with shared_examples

Malaber committed:
Overwrite owner of tournament manually

Malaber committed:
Move scope of tournament creation in match_score controller test

This fixes the problem we had with the update of the match score
returning :forbidden because the @owner variable was not set properly.

Also names of the tests should now be easier to read.

Malaber committed:
Remove unused Variable stage

Malaber committed:
Rename groups to prepared_groups

This is done because prepared_groups only represents hashes of groups
that still need to be generated rather then actual group objects.

Malaber committed:
Test update_group_scores method

Thor77 commented on app/controllers/match_scores_controller.rb:
```diff
@@ -13,6 +13,10 @@ def show
   def update
     if @match_score.update(match_score_params)
+      if @match_score.part_of_group_match?
+        match_score_group = @match_score.match.group
```
:thinking:

Malaber replied:
I apparently had bigger plans for this variable :D removed it :D

Thor77 commented on app/models/match.rb:
```diff
@@ -19,17 +19,48 @@ def owner
     match_scores.max_by(&:points).team
   end
 
+  def winner
```
```ruby
return finished? ? current_leading_team : nil
```

Malaber replied:
done

Thor77 commented on app/models/match.rb:
```diff
@@ -19,17 +19,48 @@ def owner
     group.present?
   end
 
+  def scored_points_of(team)
```
```ruby
return teams.include?(team) ? match_scores.find_by(team: team).points : 0
```

Malaber replied:
done

Thor77 commented on app/models/match.rb:
```diff
@@ -19,17 +19,48 @@ def owner
+    match_scores.find_by(team: team).points
+  end
+
+  def received_points_of(team)
```
```ruby
return teams.include?(team) ? match_scores.find { |ms| ms.team != team }.points : 0
```

Malaber replied:
I don't think that adds to readability

Malaber replied:
done

Thor77 commented on app/models/match.rb:
```diff
@@ -19,17 +19,48 @@ def owner
+
+  def group_points_of(team)
+    return 0 unless finished? || in_progress?
+    return 0 unless teams.include?(team)
```
please combine those statements

Malaber replied:
done

Thor77 commented on app/interactors/update_groups_group_scores.rb:
```diff
@@ -0,0 +1,15 @@
+  def call
+    group = context.group
+    begin
+      group_scores = GroupStageService.update_group_scores(group)
```
:thinking:

Malaber replied:
removed

Thor77 commented on app/services/group_stage_service.rb:
```diff
@@ -1,32 +1,59 @@
-      matches << match
+    # Updates all group_scores of the given group
+    #
+    # @param group Group the group to update the group_scores in
```
in?

Malaber replied:
that is an actual sentence
in german it would be:
"Die gruppe in welcher die group_scores geupdatet werden"

Thor77 replied:
too much group in there imo

Thor77 commented on spec/models/match_score_spec.rb:
```diff
@@ -7,4 +7,14 @@
+
+  describe '#part_of_group_match?' do
+    it 'is part of a group match' do
+      expect(create(:running_group_match).match_scores.first.part_of_group_match?).to be(true)
```
```suggestion
      expect(create(:running_group_match).match_scores.first.part_of_group_match?).to eq(true)
```

Malaber replied:
done

Thor77 commented on spec/models/match_score_spec.rb:
```diff
@@ -7,4 +7,14 @@
+    end
+
+    it 'isn\'t part of a group match' do
+      expect(create(:running_playoff_match).match_scores.first.part_of_group_match?).to be(false)
```
```suggestion
      expect(create(:running_playoff_match).match_scores.first.part_of_group_match?).to eq(false)
```

Malaber replied:
done

Thor77 commented on spec/services/group_stage_service_spec.rb:
```diff
@@ -72,4 +77,72 @@
+        @changed_group_scores = GroupStageService.update_group_scores(@group)
+      end
+
+      it 'assigns 1 point to every team' do
```
```suggestion
      it 'assigns one point to every team' do
```

Thor77 commented on spec/services/group_stage_service_spec.rb:
```diff
@@ -72,4 +77,72 @@
+        end
+      end
+
+      it_should_behave_like 'only_return_group_scores'
```
imo this should be a the top (or end) of this context to avoid confusion

Malaber replied:
done

Malaber committed:
Replace guard clauses with ternary operator

@thor77 apparently thinks this is better then it was before.. i don't
really care so here we go ¯\_(ツ)_/¯

Malaber commented:
Well i guess I can rephrase that.. But the first two groups are not part of the sentence and need to stay :D

> Am 05.06.2019 um 00:08 schrieb Thor77 <notifications@github.com>:
> 
> @Thor77 commented on this pull request.
> 
> In app/services/group_stage_service.rb:
> 
> >  
> -  def self.generate_all_matches_between(teams)
> -    matches = []
> -    teams.combination(2).to_a # = matchups
> -         .each_with_index do |matchup, i|
> -      match = Match.new state: :not_started,
> -                        position: i,
> -                        match_scores: [
> -                          MatchScore.new(team: matchup.first),
> -                          MatchScore.new(team: matchup.second)
> -                        ]
> -      matches << match
> +    # Updates all group_scores of the given group
> +    #
> +    # @param group Group the group to update the group_scores in
> too much group in there imo
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

Malaber committed:
Test returning unprocessable entity when match_score update fails
